src/dashboard/src/settings/base.py

```diff
@@ -501,6 +507,20 @@
 
     from .components.shibboleth_auth import *  # noqa
 
+CAS_AUTHENTICATION = config.get("cas_authentication")
+if CAS_AUTHENTICATION:
+    ALLOW_USER_EDITS = False
+    INSTALLED_APPS += ["django_cas_ng"]
+
+    AUTHENTICATION_BACKENDS += ["components.accounts.backends.CustomCASBackend"]
+
+    # Insert CAS after the authentication middleware
+    MIDDLEWARE.insert(
+        MIDDLEWARE.index("django.contrib.auth.middleware.AuthenticationMiddleware") + 1,
+        "django_cas_ng.middleware.CASMiddleware",
+    )
+
+    from .components.cas_auth import *  # noqa
 
 LDAP_AUTHENTICATION = config.get("ldap_authentication")
 if LDAP_AUTHENTICATION:
```

Comment:
It's definitely [possible and supported to have multiple authentication backends](https://docs.djangoproject.com/en/1.11/topics/auth/customizing/#specifying-authentication-backends). The custom CAS backend is being added to `AUTHENTICATION_BACKENDS` after `django.contrib.auth.backends.ModelBackend` rather than replacing it, which works fine because Django tries each authentication backend until one works.

I'm not sure what would happen if someone enabled more than one authentication scheme in our current settings. My guess is it would depend on the combination.

If it was CAS and OIDC (there is an [unmerged PR](https://github.com/artefactual/archivematica/pull/1536) which I used as a reference for the CAS work), I suspect everything would work but the user would only be able to authenticate via CAS because the CAS middleware skips the Archivematica login screen, where OIDC is presented as an option for login next to local authentication. So I don't think there would be too much value in combining them. If we added something like OAuth as an option in the Archivematica login screen, my guess is we'd see a similar behaviour. But it's also very possible there would be errors - I'm not sure.

But presumably OIDC + OAuth + local login could work just fine, with all three presented as options via the Archivematica login screen.

I ran an experiment where I enabled both CAS and Shibboleth in my `docker-compose.yml`. I was redirected to the CAS server for login and was authenticated but then a `NoReverseMatch` exception was raised because the `shibboleth` namespace wasn't be found. It seems like we could work out these sorts of issues with different combinations of auth schemes, but for schemes like CAS and Shibboleth I think one (likely the one whose middleware is higher in the list) will take precedence anyways, which begs the question of why one would want to enable them both.

What do you think? Should we build in more safeguards here?

Reply:
Thank you, that's useful. I think so, yeah. If we don't have capacity to test all permutations, we can't guarantee it's safe nor its long-term support then I'd rather make them mutually exclusive. If the user really wanted to combine them they can always write their own settings module, but at that point I think it'd be cleat that it's outside our recommendations.

Reply:
That makes sense to me. After some further thought it's CAS in particular that doesn't play nicely with the other SSO methods (because its middleware circumvents the Archivematica login screen, where other SSO options would be presented to a user or handled within the existing form), so I've added a check in the CAS settings in `base.py` that raises an `ImproperlyConfigured` exception if Shibboleth or LDAP is enabled in addition to CAS, as well as an explanatory comment. Please let me know if you think that resolves the issue.